cpp/src/io/json/nested_json_gpu.cu

```diff
@@ -330,47 +330,115 @@ enum class dfa_symbol_group_id : uint8_t {
 
-// The i-th string representing all the characters of a symbol group
-std::array<std::string, NUM_SYMBOL_GROUPS - 1> const symbol_groups{
-  {{"{"}, {"["}, {"}"}, {"]"}, {"\""}, {"\\"}, {"\n"}}};
-
 // Transition table for the default JSON and JSON lines formats
 std::array<std::array<dfa_states, NUM_SYMBOL_GROUPS>, TT_NUM_STATES> const transition_table{
-  {/* IN_STATE          {       [       }       ]       "       \      \n    OTHER */
+  {/* IN_STATE          {       [       }       ]       "       \    <delim>  OTHER */
    /* TT_OOS    */ {{TT_OOS, TT_OOS, TT_OOS, TT_OOS, TT_STR, TT_OOS, TT_OOS, TT_OOS}},
    /* TT_STR    */ {{TT_STR, TT_STR, TT_STR, TT_STR, TT_OOS, TT_ESC, TT_STR, TT_STR}},
    /* TT_ESC    */ {{TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR}}}};
 
 // Transition table for the JSON lines format that recovers from invalid JSON lines
 std::array<std::array<dfa_states, NUM_SYMBOL_GROUPS>, TT_NUM_STATES> const
   resetting_transition_table{
-    {/* IN_STATE          {       [       }       ]       "       \      \n    OTHER */
+    {/* IN_STATE          {       [       }       ]       "       \    <delim> OTHER */
      /* TT_OOS    */ {{TT_OOS, TT_OOS, TT_OOS, TT_OOS, TT_STR, TT_OOS, TT_OOS, TT_OOS}},
      /* TT_STR    */ {{TT_STR, TT_STR, TT_STR, TT_STR, TT_OOS, TT_ESC, TT_OOS, TT_STR}},
      /* TT_ESC    */ {{TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_STR, TT_OOS, TT_STR}}}};
 
-// Translation table for the default JSON and JSON lines formats
-std::array<std::array<std::vector<char>, NUM_SYMBOL_GROUPS>, TT_NUM_STATES> const translation_table{
-  {/* IN_STATE         {      [      }      ]      "      \     \n    OTHER */
-   /* TT_OOS    */ {{{'{'}, {'['}, {'}'}, {']'}, {}, {}, {}, {}}},
-   /* TT_STR    */ {{{}, {}, {}, {}, {}, {}, {}, {}}},
-   /* TT_ESC    */ {{{}, {}, {}, {}, {}, {}, {}, {}}}}};
-
-// Translation table for the JSON lines format that recovers from invalid JSON lines
-std::array<std::array<std::vector<char>, NUM_SYMBOL_GROUPS>, TT_NUM_STATES> const
-  resetting_translation_table{
-    {/* IN_STATE         {      [      }      ]      "      \     \n    OTHER */
-     /* TT_OOS    */ {{{'{'}, {'['}, {'}'}, {']'}, {}, {}, {'\n'}, {}}},
-     /* TT_STR    */ {{{}, {}, {}, {}, {}, {}, {'\n'}, {}}},
-     /* TT_ESC    */ {{{}, {}, {}, {}, {}, {}, {'\n'}, {}}}}};
+struct SymbolToSymbolGroupId {
+  SymbolT delimiter = '\n';
+  std::array<SymbolT, NUM_SYMBOL_GROUPS - 1> symbol_groups{{'{', '[', '}', ']', '"', '\\', '\n'}};
+  CUDF_HOST_DEVICE int32_t operator()(SymbolT symbol) const
+  {
+    auto it = thrust::find(thrust::device, symbol_groups.begin(), symbol_groups.end() - 1, symbol);
+    if (it != symbol_groups.end() - 1)
+      return static_cast<int32_t>(thrust::distance(symbol_groups.begin(), it));
+    if (symbol == delimiter) return static_cast<int32_t>(dfa_symbol_group_id::DELIMITER_CHAR);
+    return static_cast<int32_t>(dfa_symbol_group_id::OTHER_SYMBOLS);
+  }
+};
```

Comment:
Couldn't we just keep the original `symbol_groups` definition, remove its `const` qualifier in the declaration and just set the string of characters that map to `DELIMITER_CHAR` at runtime (e.g., right before constructing the FST), like:
```
symbol_groups[static_cast<std::uint8_t>(dfa_symbol_group_id::DELIMITER_CHAR)] = std::string{1, delimiter};
```

Reply:
Thank you for the feedback! I did consider keeping the translation tables and the symbol groups as-is, and directly modifying them in `get_stack_context` depending on the delimiter character. My concern with this was that if the lookup tables change, then we need to carefully update the code in multiple locations. 
What do you think about this idea - instead of modifying the tables and symbol groups in `get_stack_context`, we can have an update function in the same namespace which can be called if the delimiter needs to be changed?

Reply:
Sounds good. I suggest that instead of an update function, you just implement a get_sgid_lut(delim) function that constructs and returns the lookup table. We do something similar on the transition and translation tables of the full-PDA FST, just instead of a delimiter, depending on some other parser options.